### PR TITLE
Create Cloud function afterLiveQueryEvent

### PR DESCRIPTION
**Pull request by dblythy**
This is a new Cloud function that allows mutation of the `original` and `current` objects that get sent to the LiveQuery client. This is a LiveQuery equivalent of `afterFind`.

Implementation:

```
// This will change the values for "foo" and "yolo" only for LQ events
Parse.Cloud.afterLiveQueryEvent('TestObject', req => {
      const current = req.object;
      current.set('foo', 'yolo');

      const original = req.original;
      original.set('yolo', 'foo');
});

subscription.on('update', (object, original) => {
// without the trigger, object.get("foo") and original.get("yolo") would be empty
});
```

```
// This will prevent LiveQuery messages being passed to the client if foo is changed
Parse.Cloud.afterLiveQueryEvent('TestObject', req => {

      const current = req.object;
      const original = req.original;

      if (current.get('foo') != original.get('foo')) {
        throw "Don't pass an update trigger, or message";
      }
});

subscription.on('update', (object, original) => {
// this won't fire because the "throw" in the trigger, even if an object was updated.
});
```

This could also be a solution to #2946, by allowing secondary querying in the afterLiveQueryEvent block.
```
    const object = new TestObject();
    const child = new Parse.Object('Foo');
    child.set('foo','bar');
    await child.save();

    const relation = object.relation('children');
    relation.add(child);
    const query = new Parse.Query('TestObject');


    Parse.Cloud.afterLiveQueryEvent('TestObject', async req => {
      if (req.event != 'Create') {
        return;
      }
      let query = req.object.relation('children').query();
      query.equalTo('foo','bart');
      let first = await query.first();
      if (!first)  {
        throw "Doesn't match secondary query."
      }
    });


    const subscription = await query.subscribe();
    subscription.on('create', object => {
      // this won't be triggered now as children relation doesn't contain any objects with 'foo' equal to 'bar'.
    });
    await object.save();

```
Using async will obviously slow down the speed of LiveQuery events, and will require proper indexing / query design. It will also increase the required resources.

I've added a number of test cases to ensure it works as expected for "Create", "Update", "Enter", "Leave", and "Delete" events.

This feature is inspired by a few different threads.

#1686 - include in LiveQuery. This is not a direct solution, however it allows custom mutations and that could emulate .include functionality.
And [this](https://github.com/parse-community/Parse-SDK-JS/issues/1184) issue in the JS SDK, which restricts LiveQuery events to specific events.

Let me know what you think!

### CODE REVIEW

**Comment by dblythy** [comment]
This is a branch that I am using myself nonetheless. I'd be curious to see if this is something that we could possibly add to the master. Working on fixing the tests.

**Comment by dblythy** [comment]
Thank you @FNPCMDs! I would say be careful performance-wise as every LiveQuery event will trigger the secondary query unless the cloud function is restrictive. Also make sure you have your secondary query indexed and efficient as possible 😊

**Review by byeze — APPROVED** [review]

**Comment by dblythy** [comment]
@dplewis @mtrezza do you guys have any feedback / suggestions for implementing this feature so I can close this PR? Or would this feature put too much pressure on the LQ server? Cheers

**Comment by dplewis** [comment]
@dblythy Sorry for the late reply. The PR looks good. My question is should this hook return an object? Similar logic to beforeSubscribe that you wrote https://github.com/parse-community/parse-server/pull/6868#discussion_r473851402

**Comment by dblythy** [comment]
Thanks for the feedback @dplewis. I've removed the option to return an object, and also made sure that mutating req.object / req.original changes the class name 😊

**Review by dplewis — APPROVED** [review]
LGTM! I look forward to using this in production

File: src/cloud-code/Parse.Cloud.js [hunk 1]
```diff
@@ -511,6 +511,16 @@ ParseCloud.onLiveQueryEvent = function (handler) {
   triggers.addLiveQueryEventHandler(handler, Parse.applicationId);
 };
 
+ParseCloud.afterLiveQueryEvent = function (parseClass, handler) {
```
Comment (dplewis) [inline src/cloud-code/Parse.Cloud.js:514]: Can you add documentation?

**Review by dplewis — CHANGES_REQUESTED** [review]

File: src/cloud-code/Parse.Cloud.js (hunk 1 above)
Comment (dblythy) [reply src/cloud-code/Parse.Cloud.js:514]: Can do. I will also add to the docs repo once merged.

**Comment by dblythy** [comment]
Let me know if I've made a mistake in the docs formatting @dplewis. I've also added masterKey, installationId, clients, and subscription to the cloud function request for completeness. Thank you 🙏

**Review by dplewis — APPROVED** [review]

**Pull request merged**